Haiku (x86_64) port of Node.js v18.20.8, all dependencies bundled.
Made by node-haiku/scripts/port-haiku.py (+ c-ares' configure result on Haiku); apply with patch -p1.

--- a/common.gypi
+++ b/common.gypi
@@ -393,11 +393,11 @@
           'BUILDING_UV_SHARED=1',
         ],
       }],
-      [ 'OS in "linux freebsd openbsd solaris aix os400"', {
+      [ 'OS in "linux freebsd openbsd solaris aix os400 haiku"', {
         'cflags': [ '-pthread' ],
         'ldflags': [ '-pthread' ],
       }],
-      [ 'OS in "linux freebsd openbsd solaris android aix os400 cloudabi"', {
+      [ 'OS in "linux freebsd openbsd solaris android aix os400 cloudabi haiku"', {
         'cflags': [ '-Wall', '-Wextra', '-Wno-unused-parameter', ],
         'cflags_cc': [ '-fno-rtti', '-fno-exceptions', '-std=gnu++17' ],
         'defines': [ '__STDC_FORMAT_MACROS' ],
--- a/configure.py
+++ b/configure.py
@@ -46,7 +46,7 @@
 parser = argparse.ArgumentParser()
 
 valid_os = ('win', 'mac', 'solaris', 'freebsd', 'openbsd', 'linux',
-            'android', 'aix', 'cloudabi', 'os400', 'ios')
+            'android', 'aix', 'cloudabi', 'os400', 'ios', 'haiku')
 valid_arch = ('arm', 'arm64', 'ia32', 'mips', 'mipsel', 'mips64el', 'ppc',
               'ppc64', 'x64', 'x86', 'x86_64', 's390x', 'riscv64', 'loong64')
 valid_arm_float_abi = ('soft', 'softfp', 'hard')
--- a/deps/cares/cares.gyp
+++ b/deps/cares/cares.gyp
@@ -226,6 +226,13 @@
           'include_dirs': [ 'config/openbsd' ],
           'sources': [ 'config/openbsd/ares_config.h' ]
         }],
+        [ 'OS=="haiku"', {
+          'include_dirs': [ 'config/haiku' ],
+          'sources': [ 'config/haiku/ares_config.h' ],
+          'direct_dependent_settings': {
+            'libraries': [ '-lnetwork' ],
+          },
+        }],
         [ 'OS=="android"', {
           'include_dirs': [ 'config/android' ],
           'sources': [ 'config/android/ares_config.h' ],
--- /dev/null
+++ b/deps/cares/config/haiku/ares_config.h
@@ -0,0 +1,481 @@
+/* Copyright (C) The c-ares project and its contributors
+ * SPDX-License-Identifier: MIT
+ */
+
+/* Generated from ares_config.h.cmake */
+
+/* Define if building universal (internal helper macro) */
+#undef AC_APPLE_UNIVERSAL_BUILD
+
+/* Defined for build with symbol hiding. */
+/* #undef CARES_SYMBOL_HIDING */
+
+/* Use resolver library to configure cares */
+/* #undef CARES_USE_LIBRESOLV */
+
+/* if a /etc/inet dir is being used */
+#undef ETC_INET
+
+/* Define to the type of arg 2 for gethostname. */
+#define GETHOSTNAME_TYPE_ARG2 size_t
+
+/* Define to the type qualifier of arg 1 for getnameinfo. */
+#define GETNAMEINFO_QUAL_ARG1 
+
+/* Define to the type of arg 1 for getnameinfo. */
+#define GETNAMEINFO_TYPE_ARG1 struct sockaddr *
+
+/* Define to the type of arg 2 for getnameinfo. */
+#define GETNAMEINFO_TYPE_ARG2 socklen_t
+
+/* Define to the type of args 4 and 6 for getnameinfo. */
+#define GETNAMEINFO_TYPE_ARG46 socklen_t
+
+/* Define to the type of arg 7 for getnameinfo. */
+#define GETNAMEINFO_TYPE_ARG7 int
+
+/* Specifies the number of arguments to getservbyport_r */
+#define GETSERVBYPORT_R_ARGS 
+
+/* Specifies the number of arguments to getservbyname_r */
+#define GETSERVBYNAME_R_ARGS 
+
+/* Define to 1 if you have AF_INET6. */
+#define HAVE_AF_INET6
+
+/* Define to 1 if you have the <arpa/inet.h> header file. */
+#define HAVE_ARPA_INET_H
+
+/* Define to 1 if you have the <arpa/nameser_compat.h> header file. */
+#define HAVE_ARPA_NAMESER_COMPAT_H
+
+/* Define to 1 if you have the <arpa/nameser.h> header file. */
+#define HAVE_ARPA_NAMESER_H
+
+/* Define to 1 if you have the <assert.h> header file. */
+#define HAVE_ASSERT_H
+
+/* Define to 1 if you have the clock_gettime function and monotonic timer. */
+#define HAVE_CLOCK_GETTIME_MONOTONIC
+
+/* Define to 1 if you have the closesocket function. */
+/* #undef HAVE_CLOSESOCKET */
+
+/* Define to 1 if you have the CloseSocket camel case function. */
+/* #undef HAVE_CLOSESOCKET_CAMEL */
+
+/* Define to 1 if you have the connect function. */
+#define HAVE_CONNECT
+
+/* define if the compiler supports basic C++11 syntax */
+/* #undef HAVE_CXX11 */
+
+/* Define to 1 if you have the <dlfcn.h> header file. */
+#define HAVE_DLFCN_H
+
+/* Define to 1 if you have the <errno.h> header file. */
+#define HAVE_ERRNO_H
+
+/* Define to 1 if you have the <poll.h> header file. */
+#define HAVE_POLL_H
+
+/* Define to 1 if you have the poll function. */
+#define HAVE_POLL
+
+/* Define to 1 if you have the pipe function. */
+#define HAVE_PIPE
+
+/* Define to 1 if you have the pipe2 function. */
+#define HAVE_PIPE2
+
+/* Define to 1 if you have the kqueue function. */
+/* #undef HAVE_KQUEUE */
+
+/* Define to 1 if you have the epoll{_create,ctl,wait} functions. */
+/* #undef HAVE_EPOLL */
+
+/* Define to 1 if you have the fcntl function. */
+#define HAVE_FCNTL
+
+/* Define to 1 if you have the <fcntl.h> header file. */
+#define HAVE_FCNTL_H
+
+/* Define to 1 if you have a working fcntl O_NONBLOCK function. */
+#define HAVE_FCNTL_O_NONBLOCK
+
+/* Define to 1 if you have the freeaddrinfo function. */
+#define HAVE_FREEADDRINFO
+
+/* Define to 1 if you have a working getaddrinfo function. */
+#define HAVE_GETADDRINFO
+
+/* Define to 1 if the getaddrinfo function is threadsafe. */
+/* #undef HAVE_GETADDRINFO_THREADSAFE */
+
+/* Define to 1 if you have the getenv function. */
+#define HAVE_GETENV
+
+/* Define to 1 if you have the gethostname function. */
+#define HAVE_GETHOSTNAME
+
+/* Define to 1 if you have the getnameinfo function. */
+#define HAVE_GETNAMEINFO
+
+/* Define to 1 if you have the getrandom function. */
+/* #undef HAVE_GETRANDOM */
+
+/* Define to 1 if you have the getservbyport_r function. */
+/* #undef HAVE_GETSERVBYPORT_R */
+
+/* Define to 1 if you have the getservbyname_r function. */
+/* #undef HAVE_GETSERVBYNAME_R */
+
+/* Define to 1 if you have the `gettimeofday' function. */
+#define HAVE_GETTIMEOFDAY
+
+/* Define to 1 if you have the `if_indextoname' function. */
+#define HAVE_IF_INDEXTONAME
+
+/* Define to 1 if you have the `if_nametoindex' function. */
+#define HAVE_IF_NAMETOINDEX
+
+/* Define to 1 if you have the `ConvertInterfaceIndexToLuid' function. */
+/* #undef HAVE_CONVERTINTERFACEINDEXTOLUID */
+
+/* Define to 1 if you have the `ConvertInterfaceLuidToNameA' function. */
+/* #undef HAVE_CONVERTINTERFACELUIDTONAMEA */
+
+/* Define to 1 if you have a IPv6 capable working inet_net_pton function. */
+#define HAVE_INET_NET_PTON
+
+/* Define to 1 if you have a IPv6 capable working inet_ntop function. */
+#define HAVE_INET_NTOP
+
+/* Define to 1 if you have a IPv6 capable working inet_pton function. */
+#define HAVE_INET_PTON
+
+/* Define to 1 if you have the <inttypes.h> header file. */
+#define HAVE_INTTYPES_H
+
+/* Define to 1 if you have the ioctl function. */
+#define HAVE_IOCTL
+
+/* Define to 1 if you have the ioctlsocket function. */
+/* #undef HAVE_IOCTLSOCKET */
+
+/* Define to 1 if you have the IoctlSocket camel case function. */
+/* #undef HAVE_IOCTLSOCKET_CAMEL */
+
+/* Define to 1 if you have a working IoctlSocket camel case FIONBIO function.
+   */
+/* #undef HAVE_IOCTLSOCKET_CAMEL_FIONBIO */
+
+/* Define to 1 if you have a working ioctlsocket FIONBIO function. */
+/* #undef HAVE_IOCTLSOCKET_FIONBIO */
+
+/* Define to 1 if you have a working ioctl FIONBIO function. */
+#define HAVE_IOCTL_FIONBIO
+
+/* Define to 1 if you have a working ioctl SIOCGIFADDR function. */
+#define HAVE_IOCTL_SIOCGIFADDR
+
+/* Define to 1 if you have the `resolve' library (-lresolve). */
+/* #undef HAVE_LIBRESOLV */
+
+/* Define to 1 if you have iphlpapi.h */
+/* #undef HAVE_IPHLPAPI_H */
+
+/* Define to 1 if you have netioapi.h */
+/* #undef HAVE_NETIOAPI_H */
+
+/* Define to 1 if you have the <limits.h> header file. */
+#define HAVE_LIMITS_H
+
+/* Define to 1 if the compiler supports the 'long long' data type. */
+#define HAVE_LONGLONG
+
+/* Define to 1 if you have the malloc.h header file. */
+#define HAVE_MALLOC_H
+
+/* Define to 1 if you have the memory.h header file. */
+#define HAVE_MEMORY_H
+
+/* Define to 1 if you have the AvailabilityMacros.h header file. */
+/* #undef HAVE_AVAILABILITYMACROS_H */
+
+/* Define to 1 if you have the MSG_NOSIGNAL flag. */
+#define HAVE_MSG_NOSIGNAL
+
+/* Define to 1 if you have the <netdb.h> header file. */
+#define HAVE_NETDB_H
+
+/* Define to 1 if you have the <netinet/in.h> header file. */
+#define HAVE_NETINET_IN_H
+
+/* Define to 1 if you have the <netinet6/in6.h> header file. */
+#define HAVE_NETINET6_IN6_H
+
+/* Define to 1 if you have the <netinet/tcp.h> header file. */
+#define HAVE_NETINET_TCP_H
+
+/* Define to 1 if you have the <net/if.h> header file. */
+#define HAVE_NET_IF_H
+
+/* Define to 1 if you have PF_INET6. */
+#define HAVE_PF_INET6
+
+/* Define to 1 if you have the recv function. */
+#define HAVE_RECV
+
+/* Define to 1 if you have the recvfrom function. */
+#define HAVE_RECVFROM
+
+/* Define to 1 if you have the send function. */
+#define HAVE_SEND
+
+/* Define to 1 if you have the setsockopt function. */
+#define HAVE_SETSOCKOPT
+
+/* Define to 1 if you have a working setsockopt SO_NONBLOCK function. */
+/* #undef HAVE_SETSOCKOPT_SO_NONBLOCK */
+
+/* Define to 1 if you have the <signal.h> header file. */
+#define HAVE_SIGNAL_H
+
+/* Define to 1 if your struct sockaddr_in6 has sin6_scope_id. */
+#define HAVE_STRUCT_SOCKADDR_IN6_SIN6_SCOPE_ID
+
+/* Define to 1 if you have the socket function. */
+#define HAVE_SOCKET
+
+/* Define to 1 if you have the <socket.h> header file. */
+/* #undef HAVE_SOCKET_H */
+
+/* Define to 1 if you have the <stdbool.h> header file. */
+#define HAVE_STDBOOL_H
+
+/* Define to 1 if you have the <stdint.h> header file. */
+#define HAVE_STDINT_H
+
+/* Define to 1 if you have the <stdlib.h> header file. */
+#define HAVE_STDLIB_H
+
+/* Define to 1 if you have the strcasecmp function. */
+#define HAVE_STRCASECMP
+
+/* Define to 1 if you have the strcmpi function. */
+/* #undef HAVE_STRCMPI */
+
+/* Define to 1 if you have the strdup function. */
+#define HAVE_STRDUP
+
+/* Define to 1 if you have the stricmp function. */
+/* #undef HAVE_STRICMP */
+
+/* Define to 1 if you have the <strings.h> header file. */
+#define HAVE_STRINGS_H
+
+/* Define to 1 if you have the <string.h> header file. */
+#define HAVE_STRING_H
+
+/* Define to 1 if you have the strncasecmp function. */
+#define HAVE_STRNCASECMP
+
+/* Define to 1 if you have the strncmpi function. */
+/* #undef HAVE_STRNCMPI */
+
+/* Define to 1 if you have the strnicmp function. */
+/* #undef HAVE_STRNICMP */
+
+/* Define to 1 if you have the <stropts.h> header file. */
+/* #undef HAVE_STROPTS_H */
+
+/* Define to 1 if you have struct addrinfo. */
+#define HAVE_STRUCT_ADDRINFO
+
+/* Define to 1 if you have struct in6_addr. */
+#define HAVE_STRUCT_IN6_ADDR
+
+/* Define to 1 if you have struct sockaddr_in6. */
+#define HAVE_STRUCT_SOCKADDR_IN6
+
+/* if struct sockaddr_storage is defined */
+#define HAVE_STRUCT_SOCKADDR_STORAGE
+
+/* Define to 1 if you have the timeval struct. */
+#define HAVE_STRUCT_TIMEVAL
+
+/* Define to 1 if you have the <sys/ioctl.h> header file. */
+#define HAVE_SYS_IOCTL_H
+
+/* Define to 1 if you have the <sys/param.h> header file. */
+#define HAVE_SYS_PARAM_H
+
+/* Define to 1 if you have the <sys/random.h> header file. */
+/* #undef HAVE_SYS_RANDOM_H */
+
+/* Define to 1 if you have the <sys/event.h> header file. */
+#define HAVE_SYS_EVENT_H
+
+/* Define to 1 if you have the <sys/epoll.h> header file. */
+/* #undef HAVE_SYS_EPOLL_H */
+
+/* Define to 1 if you have the <sys/select.h> header file. */
+#define HAVE_SYS_SELECT_H
+
+/* Define to 1 if you have the <sys/socket.h> header file. */
+#define HAVE_SYS_SOCKET_H
+
+/* Define to 1 if you have the <sys/stat.h> header file. */
+#define HAVE_SYS_STAT_H
+
+/* Define to 1 if you have the <sys/time.h> header file. */
+#define HAVE_SYS_TIME_H
+
+/* Define to 1 if you have the <sys/types.h> header file. */
+#define HAVE_SYS_TYPES_H
+
+/* Define to 1 if you have the <sys/uio.h> header file. */
+#define HAVE_SYS_UIO_H
+
+/* Define to 1 if you have the <time.h> header file. */
+#define HAVE_TIME_H
+
+/* Define to 1 if you have the <ifaddrs.h> header file. */
+#define HAVE_IFADDRS_H
+
+/* Define to 1 if you have the <unistd.h> header file. */
+#define HAVE_UNISTD_H
+
+/* Define to 1 if you have the windows.h header file. */
+/* #undef HAVE_WINDOWS_H */
+
+/* Define to 1 if you have the winsock2.h header file. */
+/* #undef HAVE_WINSOCK2_H */
+
+/* Define to 1 if you have the winsock.h header file. */
+/* #undef HAVE_WINSOCK_H */
+
+/* Define to 1 if you have the mswsock.h header file. */
+/* #undef HAVE_MSWSOCK_H */
+
+/* Define to 1 if you have the winternl.h header file. */
+/* #undef HAVE_WINTERNL_H */
+
+/* Define to 1 if you have the ntstatus.h header file. */
+/* #undef HAVE_NTSTATUS_H */
+
+/* Define to 1 if you have the ntdef.h header file. */
+/* #undef HAVE_NTDEF_H */
+
+/* Define to 1 if you have the writev function. */
+#define HAVE_WRITEV
+
+/* Define to 1 if you have the ws2tcpip.h header file. */
+/* #undef HAVE_WS2TCPIP_H */
+
+/* Define to 1 if you have the __system_property_get function */
+/* #undef HAVE___SYSTEM_PROPERTY_GET */
+
+/* Define to 1 if you need the malloc.h header file even with stdlib.h */
+/* #undef NEED_MALLOC_H */
+
+/* Define to 1 if you need the memory.h header file even with stdlib.h */
+/* #undef NEED_MEMORY_H */
+
+/* Define if have arc4random_buf() */
+/* #undef HAVE_ARC4RANDOM_BUF */
+
+/* Define if have getifaddrs() */
+#define HAVE_GETIFADDRS
+
+/* Define if have stat() */
+#define HAVE_STAT
+
+/* a suitable file/device to read random data from */
+#define CARES_RANDOM_FILE "/dev/urandom"
+
+/* Define to the type qualifier pointed by arg 5 for recvfrom. */
+#define RECVFROM_QUAL_ARG5 
+
+/* Define to the type of arg 1 for recvfrom. */
+#define RECVFROM_TYPE_ARG1 int
+
+/* Define to the type pointed by arg 2 for recvfrom. */
+#define RECVFROM_TYPE_ARG2 void *
+
+/* Define to 1 if the type pointed by arg 2 for recvfrom is void. */
+#define RECVFROM_TYPE_ARG2_IS_VOID 0
+
+/* Define to the type of arg 3 for recvfrom. */
+#define RECVFROM_TYPE_ARG3 size_t
+
+/* Define to the type of arg 4 for recvfrom. */
+#define RECVFROM_TYPE_ARG4 int
+
+/* Define to the type pointed by arg 5 for recvfrom. */
+#define RECVFROM_TYPE_ARG5 struct sockaddr *
+
+/* Define to 1 if the type pointed by arg 5 for recvfrom is void. */
+#define RECVFROM_TYPE_ARG5_IS_VOID 0
+
+/* Define to the type pointed by arg 6 for recvfrom. */
+#define RECVFROM_TYPE_ARG6 socklen_t *
+
+/* Define to 1 if the type pointed by arg 6 for recvfrom is void. */
+#define RECVFROM_TYPE_ARG6_IS_VOID 0
+
+/* Define to the function return type for recvfrom. */
+#define RECVFROM_TYPE_RETV ssize_t
+
+/* Define to the type of arg 1 for recv. */
+#define RECV_TYPE_ARG1 int
+
+/* Define to the type of arg 2 for recv. */
+#define RECV_TYPE_ARG2 void *
+
+/* Define to the type of arg 3 for recv. */
+#define RECV_TYPE_ARG3 size_t
+
+/* Define to the type of arg 4 for recv. */
+#define RECV_TYPE_ARG4 int
+
+/* Define to the function return type for recv. */
+#define RECV_TYPE_RETV ssize_t
+
+/* Define to the type qualifier of arg 2 for send. */
+#define SEND_QUAL_ARG2 
+
+/* Define to the type of arg 1 for send. */
+#define SEND_TYPE_ARG1 int
+
+/* Define to the type of arg 2 for send. */
+#define SEND_TYPE_ARG2 void *
+
+/* Define to the type of arg 3 for send. */
+#define SEND_TYPE_ARG3 size_t
+
+/* Define to the type of arg 4 for send. */
+#define SEND_TYPE_ARG4 int
+
+/* Define to the function return type for send. */
+#define SEND_TYPE_RETV ssize_t
+
+/* Define to disable non-blocking sockets. */
+#undef USE_BLOCKING_SOCKETS
+
+/* Define to avoid automatic inclusion of winsock.h */
+#undef WIN32_LEAN_AND_MEAN
+
+/* Define to 1 if you have the pthread.h header file. */
+#define HAVE_PTHREAD_H
+
+/* Define to 1 if you have the pthread_np.h header file. */
+/* #undef HAVE_PTHREAD_NP_H */
+
+/* Define to 1 if threads are enabled */
+#define CARES_THREADS
+
+/* Define to 1 if pthread_init() exists */
+/* #undef HAVE_PTHREAD_INIT */
+
--- a/deps/icu-small/source/common/putil.cpp
+++ b/deps/icu-small/source/common/putil.cpp
@@ -1108,6 +1108,10 @@
 #endif
 }
 
+#if U_PLATFORM == U_PF_HAIKU
+extern "C" int32_t _kern_get_timezone(int32_t* offset, char* name, size_t length);
+#endif
+
 U_CAPI const char* U_EXPORT2
 uprv_tzname(int n)
 {
@@ -1163,6 +1167,17 @@
     /* else U_TZNAME will give a better result. */
 #endif
 
+#if U_PLATFORM == U_PF_HAIKU
+    /* The zone chosen in Haiku's Time preferences, as the kernel keeps it. */
+    {
+        static char haikuTimeZone[64];
+        if (_kern_get_timezone(nullptr, haikuTimeZone, sizeof(haikuTimeZone)) == 0
+                && isValidOlsonID(haikuTimeZone)) {
+            return haikuTimeZone;
+        }
+    }
+#endif
+
 #if defined(CHECK_LOCALTIME_LINK) && !defined(DEBUG_SKIP_LOCALTIME_LINK)
     /* Caller must handle threading issues */
     if (gTimeZoneBufferPtr == nullptr) {
--- a/deps/openssl/config/asn1_asm.h
+++ b/deps/openssl/config/asn1_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/asn1_no-asm.h
+++ b/deps/openssl/config/asn1_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/asn1t_asm.h
+++ b/deps/openssl/config/asn1t_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/asn1t_no-asm.h
+++ b/deps/openssl/config/asn1t_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/bio_asm.h
+++ b/deps/openssl/config/bio_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/bio_no-asm.h
+++ b/deps/openssl/config/bio_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/bn_conf_asm.h
+++ b/deps/openssl/config/bn_conf_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/bn_conf_no-asm.h
+++ b/deps/openssl/config/bn_conf_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/cmp_asm.h
+++ b/deps/openssl/config/cmp_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/cmp_no-asm.h
+++ b/deps/openssl/config/cmp_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/cms_asm.h
+++ b/deps/openssl/config/cms_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/cms_no-asm.h
+++ b/deps/openssl/config/cms_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/conf_asm.h
+++ b/deps/openssl/config/conf_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/configuration_asm.h
+++ b/deps/openssl/config/configuration_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/configuration_no-asm.h
+++ b/deps/openssl/config/configuration_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/conf_no-asm.h
+++ b/deps/openssl/config/conf_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/crmf_asm.h
+++ b/deps/openssl/config/crmf_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/crmf_no-asm.h
+++ b/deps/openssl/config/crmf_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/crypto_asm.h
+++ b/deps/openssl/config/crypto_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/crypto_no-asm.h
+++ b/deps/openssl/config/crypto_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/ct_asm.h
+++ b/deps/openssl/config/ct_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/ct_no-asm.h
+++ b/deps/openssl/config/ct_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/dso_conf_asm.h
+++ b/deps/openssl/config/dso_conf_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/dso_conf_no-asm.h
+++ b/deps/openssl/config/dso_conf_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/err_asm.h
+++ b/deps/openssl/config/err_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/err_no-asm.h
+++ b/deps/openssl/config/err_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/ess_asm.h
+++ b/deps/openssl/config/ess_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/ess_no-asm.h
+++ b/deps/openssl/config/ess_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/fipskey_asm.h
+++ b/deps/openssl/config/fipskey_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/fipskey_no-asm.h
+++ b/deps/openssl/config/fipskey_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/include_asm.h.tmpl
+++ b/deps/openssl/config/include_asm.h.tmpl
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/include_no-asm.h.tmpl
+++ b/deps/openssl/config/include_no-asm.h.tmpl
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/lhash_asm.h
+++ b/deps/openssl/config/lhash_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/lhash_no-asm.h
+++ b/deps/openssl/config/lhash_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/ocsp_asm.h
+++ b/deps/openssl/config/ocsp_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/ocsp_no-asm.h
+++ b/deps/openssl/config/ocsp_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/opensslv_asm.h
+++ b/deps/openssl/config/opensslv_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/opensslv_no-asm.h
+++ b/deps/openssl/config/opensslv_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/pkcs12_asm.h
+++ b/deps/openssl/config/pkcs12_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/pkcs12_no-asm.h
+++ b/deps/openssl/config/pkcs12_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/pkcs7_asm.h
+++ b/deps/openssl/config/pkcs7_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/pkcs7_no-asm.h
+++ b/deps/openssl/config/pkcs7_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/safestack_asm.h
+++ b/deps/openssl/config/safestack_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/safestack_no-asm.h
+++ b/deps/openssl/config/safestack_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/srp_asm.h
+++ b/deps/openssl/config/srp_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/srp_no-asm.h
+++ b/deps/openssl/config/srp_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/ssl_asm.h
+++ b/deps/openssl/config/ssl_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/ssl_no-asm.h
+++ b/deps/openssl/config/ssl_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/ui_asm.h
+++ b/deps/openssl/config/ui_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/ui_no-asm.h
+++ b/deps/openssl/config/ui_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/x509_asm.h
+++ b/deps/openssl/config/x509_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/x509_no-asm.h
+++ b/deps/openssl/config/x509_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/x509v3_asm.h
+++ b/deps/openssl/config/x509v3_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/x509v3_no-asm.h
+++ b/deps/openssl/config/x509v3_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/x509_vfy_asm.h
+++ b/deps/openssl/config/x509_vfy_asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/config/x509_vfy_no-asm.h
+++ b/deps/openssl/config/x509_vfy_no-asm.h
@@ -1,5 +1,5 @@
 #undef OPENSSL_LINUX
-#if defined(__linux) && !defined(__ANDROID__)
+#if (defined(__linux) && !defined(__ANDROID__)) || defined(__HAIKU__)
 # define OPENSSL_LINUX 1
 #endif
 
--- a/deps/openssl/openssl_common.gypi
+++ b/deps/openssl/openssl_common.gypi
@@ -58,6 +58,16 @@
         'ENGINESDIR="/dev/null"',
         '__EXTENSIONS__'
       ],
+    }, 'OS=="haiku"', {
+      'cflags': ['-Wno-missing-field-initializers',],
+      'defines': [
+        'OPENSSLDIR="/boot/system/data/ssl"',
+        'ENGINESDIR="/dev/null"',
+        'TERMIOS',
+      ],
+      # Haiku has no libdl (dlopen() is in libroot); sockets are in libnetwork.
+      'libraries!': ['-ldl -pthread', '-lm -ldl -pthread'],
+      'libraries': ['-lm', '-pthread', '-lnetwork'],
     }, {
       # linux and others
       'cflags': ['-Wno-missing-field-initializers',],
--- a/deps/uv/uv.gyp
+++ b/deps/uv/uv.gyp
@@ -343,6 +343,20 @@
             '_THREAD_SAFE',
           ],
         }],
+        [ 'OS=="haiku"', {
+          'defines': [ '_BSD_SOURCE' ],
+          'sources': [
+            'src/unix/haiku.c',
+            'src/unix/bsd-ifaddrs.c',
+            'src/unix/no-fsevents.c',
+            'src/unix/no-proctitle.c',
+            'src/unix/posix-hrtime.c',
+            'src/unix/posix-poll.c',
+          ],
+          'link_settings': {
+            'libraries': [ '-lbsd', '-lnetwork' ],
+          },
+        }],
         [ 'OS=="freebsd" or OS=="dragonflybsd"', {
           'sources': [ 'src/unix/freebsd.c' ],
         }],
--- a/deps/v8/include/v8config.h
+++ b/deps/v8/include/v8config.h
@@ -69,6 +69,7 @@
 //  V8_OS_CYGWIN        - Cygwin
 //  V8_OS_DRAGONFLYBSD  - DragonFlyBSD
 //  V8_OS_FREEBSD       - FreeBSD
+//  V8_OS_HAIKU         - Haiku
 //  V8_OS_FUCHSIA       - Fuchsia
 //  V8_OS_LINUX         - Linux (Android, ChromeOS, Linux, ...)
 //  V8_OS_DARWIN        - Darwin (macOS, iOS)
@@ -131,6 +132,11 @@
 # define V8_OS_POSIX 1
 # define V8_OS_STRING "freebsd"
 
+#elif defined(__HAIKU__)
+# define V8_OS_HAIKU 1
+# define V8_OS_POSIX 1
+# define V8_OS_STRING "haiku"
+
 #elif defined(__Fuchsia__)
 # define V8_OS_FUCHSIA 1
 # define V8_OS_POSIX 1
--- a/deps/v8/src/base/export-template.h
+++ b/deps/v8/src/base/export-template.h
@@ -83,6 +83,7 @@
 // definition sites instead.
 #define EXPORT_TEMPLATE_DECLARE_MSVC_HACK(export, _)
 #define EXPORT_TEMPLATE_DEFINE_MSVC_HACK(export, _) export
+#define EXPORT_TEMPLATE_TEST_MSVC_HACK_DEFAULT(...) true
 
 // EXPORT_TEMPLATE_STYLE is an internal helper macro that identifies which
 // export style needs to be used for the provided FOO_EXPORT macro definition.
@@ -159,5 +160,6 @@
 #undef EXPORT_TEMPLATE_TEST
 #undef EXPORT_TEMPLATE_TEST_DEFAULT_DEFAULT
 #undef EXPORT_TEMPLATE_TEST_MSVC_HACK_MSVC_HACK
+#undef EXPORT_TEMPLATE_TEST_MSVC_HACK_DEFAULT
 
 #endif  // V8_BASE_EXPORT_TEMPLATE_H_
--- /dev/null
+++ b/deps/v8/src/base/platform/platform-haiku.cc
@@ -0,0 +1,92 @@
+// Copyright 2012 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Platform-specific code for Haiku goes here. For the POSIX-compatible
+// parts, the implementation is in platform-posix.cc.
+//
+// One file for the V8 versions of Node.js 16 to 26 (V8 9.4 to 14.6); the
+// few entry points whose names or signatures changed are picked by version.
+
+#include <OS.h>
+#include <image.h>
+#include <pthread.h>
+#include <semaphore.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <sys/resource.h>
+#include <sys/time.h>
+#include <sys/types.h>
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdarg.h>
+#include <strings.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <cmath>
+
+#undef MAP_TYPE
+
+#include "include/v8-version.h"
+#include "src/base/macros.h"
+#include "src/base/platform/platform-posix-time.h"
+#include "src/base/platform/platform-posix.h"
+#include "src/base/platform/platform.h"
+
+#if V8_MAJOR_VERSION > 12 || (V8_MAJOR_VERSION == 12 && V8_MINOR_VERSION >= 9)
+#include <optional>
+#endif
+
+namespace v8 {
+namespace base {
+
+TimezoneCache* OS::CreateTimezoneCache() {
+  return new PosixDefaultTimezoneCache();
+}
+
+std::vector<OS::SharedLibraryAddress> OS::GetSharedLibraryAddresses() {
+  std::vector<SharedLibraryAddress> result;
+  image_info info;
+  int32 cookie = 0;
+  while (get_next_image_info(B_CURRENT_TEAM, &cookie, &info) == B_OK) {
+    uintptr_t start = reinterpret_cast<uintptr_t>(info.text);
+    result.push_back(
+        SharedLibraryAddress(info.name, start, start + info.text_size));
+  }
+  return result;
+}
+
+void OS::SignalCodeMovingGC() {}
+
+void OS::AdjustSchedulingParams() {}
+
+#if V8_MAJOR_VERSION > 12 || (V8_MAJOR_VERSION == 12 && V8_MINOR_VERSION >= 9)
+std::optional<OS::MemoryRange> OS::GetFirstFreeMemoryRangeWithin(
+    OS::Address boundary_start, OS::Address boundary_end, size_t minimum_size,
+    size_t alignment) {
+  return std::nullopt;
+}
+#elif V8_MAJOR_VERSION >= 10
+std::vector<OS::MemoryRange> OS::GetFreeMemoryRangesWithin(
+    OS::Address boundary_start, OS::Address boundary_end, size_t minimum_size,
+    size_t alignment) {
+  return {};
+}
+#endif
+
+// The stack grows down: its start is the top, stack_end in Haiku's terms.
+#if V8_MAJOR_VERSION >= 11
+Stack::StackSlot Stack::ObtainCurrentThreadStackStart() {
+#else
+Stack::StackSlot Stack::GetStackStart() {
+#endif
+  thread_info info;
+  if (get_thread_info(find_thread(nullptr), &info) != B_OK) return nullptr;
+  return info.stack_end;
+}
+
+}  // namespace base
+}  // namespace v8
--- a/deps/v8/src/base/platform/platform-posix.cc
+++ b/deps/v8/src/base/platform/platform-posix.cc
@@ -64,7 +64,7 @@
 #include <sys/resource.h>
 #endif
 
-#if !defined(_AIX) && !defined(V8_OS_FUCHSIA)
+#if !defined(_AIX) && !defined(V8_OS_FUCHSIA) && !defined(__HAIKU__)
 #include <sys/syscall.h>
 #endif
 
@@ -1219,7 +1219,7 @@
 // keep this version in POSIX as most Linux-compatible derivatives will
 // support it. MacOS and FreeBSD are different here.
 #if !defined(V8_OS_FREEBSD) && !defined(V8_OS_DARWIN) && !defined(_AIX) && \
-    !defined(V8_OS_SOLARIS)
+    !defined(V8_OS_SOLARIS) && !defined(V8_OS_HAIKU)
 
 // static
 Stack::StackSlot Stack::GetStackStart() {
--- a/deps/v8/src/libsampler/sampler.cc
+++ b/deps/v8/src/libsampler/sampler.cc
@@ -15,7 +15,7 @@
 #include <sys/time.h>
 #include <atomic>
 
-#if !V8_OS_QNX && !V8_OS_AIX
+#if !V8_OS_QNX && !V8_OS_AIX && !V8_OS_HAIKU
 #include <sys/syscall.h>
 #endif
 
@@ -23,7 +23,7 @@
 #include <mach/mach.h>
 // OpenBSD doesn't have <ucontext.h>. ucontext_t lives in <signal.h>
 // and is a typedef for struct sigcontext. There is no uc_mcontext.
-#elif !V8_OS_OPENBSD
+#elif !V8_OS_OPENBSD && !V8_OS_HAIKU
 #include <ucontext.h>
 #endif
 
@@ -498,6 +498,12 @@
   state->sp = reinterpret_cast<void*>(mcontext.__gregs[_REG_SP]);
   state->fp = reinterpret_cast<void*>(mcontext.__gregs[_REG_FP]);
 #endif  // V8_HOST_ARCH_*
+#elif V8_OS_HAIKU
+#if V8_HOST_ARCH_X64
+  state->pc = reinterpret_cast<void*>(mcontext.rip);
+  state->sp = reinterpret_cast<void*>(mcontext.rsp);
+  state->fp = reinterpret_cast<void*>(mcontext.rbp);
+#endif  // V8_HOST_ARCH_*
 #elif V8_OS_NETBSD
 #if V8_HOST_ARCH_IA32
   state->pc = reinterpret_cast<void*>(mcontext.__gregs[_REG_EIP]);
--- a/deps/v8/src/trap-handler/trap-handler.h
+++ b/deps/v8/src/trap-handler/trap-handler.h
@@ -119,7 +119,7 @@
   return g_is_trap_handler_enabled;
 }
 
-#if defined(V8_OS_AIX)
+#if defined(V8_OS_AIX) || defined(V8_OS_HAIKU)
 // `thread_local` does not link on AIX:
 // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100641
 extern __thread int g_thread_in_wasm_code;
--- a/node.gypi
+++ b/node.gypi
@@ -266,6 +266,18 @@
         }],
       ],
     }],
+    [ 'OS=="haiku"', {
+      'libraries': [
+        '-lbsd',
+        '-lnetwork',
+      ],
+      # libroot's locale backend loads Haiku's own ICU into the process;
+      # keep the bundled ICU out of the dynamic symbol table so the two
+      # copies cannot bind to each other (same-version names clash).
+      'ldflags': [
+        '-Wl,--exclude-libs,libicudata.a:libicui18n.a:libicuucx.a',
+      ],
+    }],
     [ 'OS=="solaris"', {
       'libraries': [
         '-lkstat',
@@ -280,7 +292,7 @@
         'NODE_PLATFORM="sunos"',
       ],
     }],
-    [ '(OS=="freebsd" or OS=="linux") and node_shared=="false"'
+    [ '(OS=="freebsd" or OS=="linux" or OS=="haiku") and node_shared=="false"'
         ' and force_load=="true"', {
       'ldflags': [
         '-Wl,-z,noexecstack',
@@ -374,7 +386,7 @@
                 },
               },
               'conditions': [
-                ['OS in "linux freebsd" and node_shared=="false"', {
+                ['OS in "linux freebsd haiku" and node_shared=="false"', {
                   'ldflags': [
                     '-Wl,--whole-archive,'
                       '<(obj_dir)/deps/openssl/<(openssl_product)',
--- a/src/debug_utils.cc
+++ b/src/debug_utils.cc
@@ -14,7 +14,7 @@
 
 #if defined(__linux__) && !defined(__GLIBC__) || \
     defined(__UCLIBC__) || \
-    defined(_AIX)
+    defined(_AIX) || defined(__HAIKU__)
 #define HAVE_EXECINFO_H 0
 #else
 #define HAVE_EXECINFO_H 1
--- a/src/node.cc
+++ b/src/node.cc
@@ -465,6 +465,11 @@
   for (unsigned nr = 1; nr < kMaxSignal; nr += 1) {
     if (nr == SIGKILL || nr == SIGSTOP)
       continue;
+#ifdef __HAIKU__
+    // SIGKILLTHR cannot be caught or ignored either.
+    if (nr == SIGKILLTHR)
+      continue;
+#endif
     act.sa_handler = (nr == SIGPIPE || nr == SIGXFSZ) ? SIG_IGN : SIG_DFL;
     if (act.sa_handler == SIG_DFL) {
       // The only bad handler value we can inhert from before exec is SIG_IGN
--- a/src/node_report.cc
+++ b/src/node_report.cc
@@ -739,16 +739,16 @@
     {"core_file_size_blocks", RLIMIT_CORE},
     {"data_seg_size_kbytes", RLIMIT_DATA},
     {"file_size_blocks", RLIMIT_FSIZE},
-#if !(defined(_AIX) || defined(__sun))
+#if !(defined(_AIX) || defined(__sun) || defined(__HAIKU__))
     {"max_locked_memory_bytes", RLIMIT_MEMLOCK},
 #endif
-#ifndef __sun
+#if !(defined(__sun) || defined(__HAIKU__))
     {"max_memory_size_kbytes", RLIMIT_RSS},
 #endif
     {"open_files", RLIMIT_NOFILE},
     {"stack_size_bytes", RLIMIT_STACK},
     {"cpu_time_seconds", RLIMIT_CPU},
-#ifndef __sun
+#if !(defined(__sun) || defined(__HAIKU__))
     {"max_user_processes", RLIMIT_NPROC},
 #endif
 #ifndef __OpenBSD__
--- a/tools/gyp/pylib/gyp/common.py
+++ b/tools/gyp/pylib/gyp/common.py
@@ -449,6 +449,8 @@
         return "zos"
     if sys.platform == "os400":
         return "os400"
+    if sys.platform.startswith("haiku"):
+        return "haiku"
 
     return "linux"
 
--- a/tools/utils.py
+++ b/tools/utils.py
@@ -59,6 +59,8 @@
     return 'win32'
   elif id == 'FreeBSD':
     return 'freebsd'
+  elif id == 'Haiku':
+    return 'haiku'
   elif id == 'OpenBSD':
     return 'openbsd'
   elif id == 'SunOS':
--- a/tools/v8_gypfiles/features.gypi
+++ b/tools/v8_gypfiles/features.gypi
@@ -46,6 +46,11 @@
       }, {
         'is_fuchsia': 0,
       }],
+      ['OS == "haiku"', {
+        'is_haiku': 1,
+      }, {
+        'is_haiku': 0,
+      }],
       ['OS=="android"', { # GYP reverts OS to linux so use `-D OS=android`
         'is_android': 1,
       }, {
--- a/tools/v8_gypfiles/v8.gyp
+++ b/tools/v8_gypfiles/v8.gyp
@@ -1145,6 +1145,12 @@
             '<(V8_ROOT)/src/base/platform/platform-fuchsia.cc',
           ]
         }],
+        ['is_haiku', {
+          'sources': [
+            '<(V8_ROOT)/src/base/debug/stack_trace_posix.cc',
+            '<(V8_ROOT)/src/base/platform/platform-haiku.cc',
+          ]
+        }],
         ['OS == "mac" or (_toolset=="host" and host_os=="mac")', {
           'sources': [
             '<(V8_ROOT)/src/base/debug/stack_trace_posix.cc',
